### ISSUE TITLE
Same file opened twice on Windows when using URL starting with "file/:"

### ISSUE THREAD

**Issue by MaximilianKoestler** [microsoft/vscode#108660]
- VSCode Version: 1.49.2, 1.5.0
- OS Version: Windows 10

### Windows
1. Create a file (this example assumes `C:\test.txt`).
2. Open the integrated terminal in vscode.
3. Paste the following line into your terminal.
   ```
   C:\test.txt file:/C:/test.txt
   ```
4. Click on the file names/links in the terminal one after another.

Both links will open the same file, but in two different tabs. vscode will not detect that this is in fact the same file.

![image](https://user-images.githubusercontent.com/6814304/95984972-425dff00-0e24-11eb-904d-9026670bef02.png)

### Linux
I was not able to reproduce this issue on Linux using a similar approach with`/test.txt file:/test.txt`. On Linux this just opens the file once.

## Why this is bad
For me this issue is pretty annoying because my toolchain sometimes produces links starting with `file:/` and when I open them, I not only sometimes open a file twice (bad with unsaved changes), this also registers in the history causing this to happen again when using Ctrl + P.
Also my language server yells at me because suddenly I have duplicate definitions (1 in the real file and 1 in this ghost copy I just opened).

Does this issue occur when all extensions are disabled?: Yes

**Comment by bpasero** [comment]
@Tyriar back to you, I am not sure what kind of URI you are creating and passing to me, but it seems wrong to me (missing drive letter):

![image](https://user-images.githubusercontent.com/900690/97956504-23aab280-1da9-11eb-826c-18740169ac33.png)

**Comment by Tyriar** [comment]
Not reproducing in Code OSS but is in Insiders and it doesn't seem to be related to an extension 🤔

**Comment by MaximilianKoestler** [comment]
@Tyriar I can still reproduce the problem with the original description under
```
Version: 1.61.0 (system setup)
Commit: ee8c7def80afc00dd6e593ef12f37756d8f504ea
Date: 2021-10-07T18:13:09.652Z
Electron: 13.5.1
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.22000
```

**Comment by Tyriar** [comment]
Yep same, that was more of a note to self or whoever looks into the issue 🙂

**Comment by DanTup** [comment]
This [came up in Dart too](https://github.com/dart-lang/sdk/issues/47534) - it's possible to click links in the terminal that open files without their drive letters. This results in the LSP client sending them to the language server, and the language server doesn't know how to handle them (I don't think it should be making assumptions about what the correct drive letter is).

Even if it's a bug in the terminal for not detecting the full clickable path correct, isn't there also a bug that VS Code is opening an editor for an incomplete path? Shouldn't it reject that in the same way as if the path is not a valid file (it fails to open an editor and instead shows an error in the quick-pick).

**Comment by DanTup** [comment]
Might have been fixed by https://github.com/microsoft/vscode/pull/138873? Seems ok for me now:

![image](https://user-images.githubusercontent.com/1078012/153839278-6292be7c-e5a9-4745-b770-ff18cc66ccd2.png)

**Comment by DanTup** [comment]
Scratch that - seems to depend on the drive letter somehow. Running the same script from different drives shows the issue for file:///C:/ but not for file:///M:/

The difference is that whe nit occurs, you don't need to press Ctrl to activate it, so I think there are two levels of link detection going on, and one is doing only paths (not URIs) and is confused by the part of a `file:///` on Windows which looks like it contains a path (without the drive letter), and VS Codeis able to resolve it as it's relative to the "current drive".

@Tyriar if so, it seems like there may be two fixes:

- fix the link detection (that runs without needing to hold Ctrl) to handle file:/// URIs
- change it to discard "valid paths" it founds that don't have drive letters on Windows

I could look at contributing a fix if you can give me a pointer and a preference.

**Comment by Tyriar** [comment]
I believe the format in the original report `file:/C:/test.txt` is invalid. This format works and opens the proper URI:

```
file:///c:/Github/Tyriar/xterm.js/README.md
file:///c:/Github/Tyriar/xterm.js/README.md:20
```

The [hostname part of the format](https://en.wikipedia.org/wiki/File_URI_scheme#Windows) does not work, but it also doesn't work in the editor so it would be a separate feature request which may already exist.

@DanTup let me know if I'm wrong, a lot of changes happened in terminal links over the past 6 months or so.

**Comment by DanTup** [comment]
@Tyriar my issue above was not related to an invalid format, it was that VS Code was detecting the link without the driver letter. If I printed:

```
file:///C:/foo/bar.txt
file:///M:/foo/bar.txt
```

Then for one of the paths it would detect the whole link, and for the other, it would only link `/foo/bar.txt` which would then open the "duplicate" in VS Code. I'm not on Windows atm, but I'll try and get a concrete example (if it's still occurring) shortly.

**Comment by DanTup** [comment]
Although FWIW, I don't know that the issue is terminal-specific. I think when VS Code is given a drive-relative path like `/foo/bar.txt` it should be resolving it to a full with-drive-letter path in any case. There are many routes that VS Code could get a drive-letter-less path and I don't think it's ever the right thing to pretend it's a different file.
